plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergV2.java

```diff
@@ -285,6 +287,98 @@ public void testSelectivelyOptimizingLeavesEqualityDeletes()
+        Table icebergTable = updateTableToV2(tableName);
+        Assertions.assertThat(icebergTable.currentSnapshot().summary().get("total-equality-deletes")).isEqualTo("0");
+        Path metadataDir = new Path(metastoreDir.toURI());
+        TrinoFileSystem fs = HDFS_FILE_SYSTEM_FACTORY.create(SESSION);
+
+        String deleteFile1 = "delete_file_" + UUID.randomUUID();
+        List<String> firstDeleteFileColumns = ImmutableList.of("regionkey", "name");
+        Schema deleteRowSchema = icebergTable.schema().select(firstDeleteFileColumns);
+        List<Integer> equalityFieldIds = firstDeleteFileColumns.stream()
+                .map(name -> deleteRowSchema.findField(name).fieldId())
+                .collect(toImmutableList());
+        Parquet.DeleteWriteBuilder writerBuilder = Parquet.writeDeletes(new ForwardingFileIo(fs).newOutputFile(new Path(metadataDir, deleteFile1).toString()))
+                .forTable(icebergTable)
+                .rowSchema(deleteRowSchema)
+                .createWriterFunc(GenericParquetWriter::buildWriter)
+                .equalityFieldIds(equalityFieldIds)
+                .overwrite();
+        EqualityDeleteWriter<Record> writer = writerBuilder.buildEqualityWriter();
+
+        Record dataDelete = GenericRecord.create(deleteRowSchema);
+        try (Closeable ignored = writer) {
+            writer.write(dataDelete.copy(ImmutableMap.of("regionkey", 1L, "name", "BRAZIL")));
+        }
+        icebergTable.newRowDelta().addDeletes(writer.toDeleteFile()).commit();
+
+        String deleteFile2 = "delete_file_" + UUID.randomUUID();
+        List<String> secondDeleteFileColumns = ImmutableList.of("name", "regionkey");
+        Schema secondDeleteRowSchema = icebergTable.schema().select(secondDeleteFileColumns);
+        equalityFieldIds = secondDeleteFileColumns.stream()
+                .map(name -> secondDeleteRowSchema.findField(name).fieldId())
+                .collect(toImmutableList());
+        writerBuilder = Parquet.writeDeletes(new ForwardingFileIo(fs).newOutputFile(new Path(metadataDir, deleteFile2).toString()))
+                .forTable(icebergTable)
+                .rowSchema(secondDeleteRowSchema)
+                .createWriterFunc(GenericParquetWriter::buildWriter)
+                .equalityFieldIds(equalityFieldIds)
+                .overwrite();
+        writer = writerBuilder.buildEqualityWriter();
+
+        dataDelete = GenericRecord.create(secondDeleteRowSchema);
+        try (Closeable ignored = writer) {
+            writer.write(dataDelete.copy(ImmutableMap.of("regionkey", 2L, "name", "INDIA")));
+        }
+        icebergTable.newRowDelta().addDeletes(writer.toDeleteFile()).commit();
+
+        assertQuery("SELECT * FROM " + tableName, "SELECT * FROM nation WHERE NOT ((regionkey = 1 AND name = 'BRAZIL') OR (regionkey = 2 AND name = 'INDIA'))");
+        assertUpdate("DROP TABLE " + tableName);
+    }
+
```

Comment:
Please add a test doing equality deletes for nested columns.

Reply:
Do you mean deleted by nested colume?

Reply:
Yes, exactly.

Reply:
Iceberg may not support nested column as delete column, I don't know how to add it.

Reply:
```
    @Test
    public void testMultipleEqualityDeletesNestedFields()
            throws Exception
    {
        String tableName = "test_equality_delets_different_schemas_" + randomNameSuffix();
        assertUpdate("CREATE TABLE " + tableName + " ( id BIGINT, root ROW(f1 BIGINT, f2 BIGINT))");
        assertUpdate("INSERT INTO " + tableName + " VALUES (1, row(10, 100))", 1);
        assertUpdate("INSERT INTO " + tableName + " VALUES (2, row(20, 200))", 1);
        assertUpdate("INSERT INTO " + tableName + " VALUES (2, row(20, 200))", 1);
        Table icebergTable = loadTable(tableName);
        assertThat(icebergTable.currentSnapshot().summary().get("total-equality-deletes")).isEqualTo("0");
        Path metadataDir = new Path(metastoreDir.toURI());
        TrinoFileSystem fs = HDFS_FILE_SYSTEM_FACTORY.create(SESSION);

        List<String> deleteFileColumns = ImmutableList.of("root.f1");
        Schema deleteRowSchema = icebergTable.schema().select(deleteFileColumns);
        List<Integer> equalityFieldIds = deleteFileColumns.stream()
                .map(name -> deleteRowSchema.findField(name).fieldId())
                .collect(toImmutableList());
        for (int i = 1; i < 3; i++) {
            String deleteFile = "delete_file_" + UUID.randomUUID();
            Parquet.DeleteWriteBuilder writerBuilder = Parquet.writeDeletes(new ForwardingFileIo(fs).newOutputFile(new Path(metadataDir, deleteFile).toString()))
                    .forTable(icebergTable)
                    .rowSchema(deleteRowSchema)
                    .createWriterFunc(GenericParquetWriter::buildWriter)
                    .equalityFieldIds(equalityFieldIds)
                    .overwrite();
            EqualityDeleteWriter<Record> writer = writerBuilder.buildEqualityWriter();

            Types.StructType nestedStructType = (Types.StructType)deleteRowSchema.findField("root").type();
            Record dataDelete = GenericRecord.create(deleteRowSchema);
            try (Closeable ignored = writer) {

                Record nestedStruct = GenericRecord.create(nestedStructType);
                nestedStruct.setField("f1", 20L);
                writer.write(dataDelete.copy(ImmutableMap.of("root", nestedStruct)));
            }
            icebergTable.newRowDelta().addDeletes(writer.toDeleteFile()).commit();
        }

        assertQuery("SELECT * FROM " + tableName, "VALUES ( BIGINT '1', row(BIGINT '10', BIGINT '100'))");
        assertUpdate("DROP TABLE " + tableName);
    }
```

@Heltman 

the above mentioned test fails with the message

```
2023-08-09T15:16:47.338-0600	ERROR	stage-scheduler	io.trino.execution.StageStateMachine	Stage 20230809_211647_00008_s6wq4.0 failed
io.trino.spi.TrinoException: Multiple entries with same key: 3=f1 and 3=root.f1
	at io.trino.plugin.iceberg.IcebergPageSource.getNextPage(IcebergPageSource.java:142)
	at io.trino.operator.TableScanOperator.getOutput(TableScanOperator.java:299)
	at io.trino.operator.Driver.processInternal(Driver.java:395)
	at io.trino.operator.Driver.lambda$process$8(Driver.java:298)
```

Reply:
this because `requiredColumns` don't judge delete column is nested child column, we need fix this in this pr or create new pr to fix this? 
https://github.com/trinodb/trino/blob/eea73300e6490de984fad9114c8e111d76ffe75b/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java#L253-L255

Reply:
when skip this bug, we will get another error, because `io.trino.plugin.iceberg.IcebergPageSink.getIcebergValue` don't support row type until now
```log
io.trino.spi.TrinoException: Type not supported as partition column: row(f1 bigint)
	at io.trino.plugin.iceberg.IcebergPageSource.getNextPage(IcebergPageSource.java:142)
	at io.trino.operator.TableScanOperator.getOutput(TableScanOperator.java:299)
        ...
Caused by: java.lang.UnsupportedOperationException: Type not supported as partition column: row(f1 bigint)
	at io.trino.plugin.iceberg.IcebergPageSink.getIcebergValue(IcebergPageSink.java:496)
	at io.trino.plugin.iceberg.delete.TrinoRow.<init>(TrinoRow.java:35)
	at io.trino.plugin.iceberg.delete.EqualityDeleteFilter.readEqualityDeletes(EqualityDeleteFilter.java:64)
	at io.trino.plugin.iceberg.IcebergPageSourceProvider.readDeletes(IcebergPageSourceProvider.java:435)
	at io.trino.plugin.iceberg.IcebergPageSourceProvider.lambda$createPageSource$8(IcebergPageSourceProvider.java:335)
	at com.google.common.base.Suppliers$NonSerializableMemoizingSupplier.get(Suppliers.java:181)
	at io.trino.plugin.iceberg.IcebergPageSource.getNextPage(IcebergPageSource.java:125)
	... 13 more
```

Reply:
Add this test, we will fix it later